File: script/Pharo.rb

# Adapter classes for Pharo

require 'stringio'
require 'os'
require 'pathname'

#------------------------------------------------------------------------------
#
class Transcript
  def self.<< (msg)
    print msg
  end
  
  def self.cr()
    puts ''
    STDOUT.flush
  end
end

#------------------------------------------------------------------------------
#
def output(msg)
  puts msg
  STDOUT.flush
end

#------------------------------------------------------------------------------
#
class Association
  attr_accessor :key, :value

  def initialize(k, v)
    @key, @value = k, v
  end
end

#------------------------------------------------------------------------------
#
class SWString < String

  def self.cr()
    return "\r"
  end
  def self.lf()
    return "\n"
  end
  def self.crlf()
    return "\r\n"
  end

  def readStream()
    return SWStringIO.new(self)
  end

  def self.streamContents_(aBlock)
    s = StringIO.new
    aBlock.call(s)
    s.rewind
    return self.new(s.read)
  end

  def trimLeft_(procedure)
    for i in 0...self.length do
      if not procedure.call(self[i]) then
        return self[i, self.length - i]
      end
    end
    return ''
  end

end

#------------------------------------------------------------------------------
#
class SWStringIO < StringIO
  def contents()
    self.rewind
    return self.read
  end
  def upToEnd()
    return self.read
  end
end

#------------------------------------------------------------------------------
#
class SWStringIOWrapper
  def initialize(entity)
    @entity = entity
  end
  def contents()
    @entity.rewind
    return @entity.read
  end
  def upToEnd()
    return @entity.read
  end
end

#------------------------------------------------------------------------------
#
class TestCase
  def setUp()
  end

  def tearDown()
  end

  def assert_equals_(left, right)
    if left != right
      raise "Assertion failed: #{left} != #{right}."
    end
  end

  def assert_(aBlock)
    if !(aBlock.call)
      raise "Assertion failed: #{aBlock.to_s} failed."
    end
  end
end

#
class OSPlatform
  def self.current
    return self
  end

  def self.isMacOS
    return OS.mac?
  end

end

#------------------------------------------------------------------------------
#
class SWDictionary < Hash
end

#------------------------------------------------------------------------------
#
class FileTime < Time
  def prettyPrint()
    return self.to_s
  end
end

#------------------------------------------------------------------------------
#
class FileLocator < Pathname
  def self.imageDirectory()
    return FileLocator.new(Dir.getwd)
  end

  def isDirectory()
    return self.directory?
  end

  def exists()
    return self.exist?
  end

  def fullName()
    return self.to_s
  end

  def basename()
    return super.to_s
  end
  
  def creationTime()
    return FileTime.at(self.ctime.to_i)
  end

  def accessTime()
    return FileTime.at(self.atime.to_i)
  end

  def changeTime()
    return FileTime.at(self.mtime.to_i)
  end

  def relativeTo_(relBase)
    return FileLocator.new(self.relative_path_from(relBase))
  end

  def contains_(path)
    base = self.to_s
    child = path.to_s
    return (base.size < child.size and base == child[0, base.size])
  end

  def parent()
    return FileLocator.new(super)
  end

  def ensureCreateDirectory()
    if (self.exist? and self.directory?) then
      return
    end
    self.mkdir
  end

  def /(path)
    return FileLocator.new(super)
  end

  def writeStreamDo_(procedure)
    io = SWFile.new(self.to_s)
    begin
      procedure.call(io)
    ensure
      io.close
    end
  end

end


#------------------------------------------------------------------------------
#
class SWFile < File
  def initialize(path)
    super(path, "a")
  end

  def truncate()
    super(self.pos)
  end

  def position_(p)
    self.pos = p
  end


end


# for Dummy
class SWGFortranCompilerAdapter
end

File: script/Pharo.mjs

//-----------------------------------------------------------------------------
export class PharoObject {

  error_(msg) {
    throw msg;
  }

  assert_(aBlock) {
    var value = false;
    if (typeof aBlock == 'function') {
      value = aBlock();
    } else {
      value = aBlock;
    }
    if(value === undefined || value === null || value === false)
        throw `Assertion failed: ${aBlock} failed.`;
  }

  static name() {
    return this.constructor.name;
  }

}
PharoObject.error_ = function(msg) {
  throw msg;
};

//-----------------------------------------------------------------------------
export class Transcript {};

Transcript.show_ = function(msg) {
  console.log(msg);
};
Transcript.write_ = function(msg) {
  process.stdout.write(msg);
};
Transcript.cr = function(msg) {
  process.stdout.write('\n');
};

//-----------------------------------------------------------------------------
export class OSPlatform {}
OSPlatform.current = function() {
  return OSPlatform;
};
OSPlatform.isMacOS = function() {
  return 'darwin' == process.platform;
};

//-----------------------------------------------------------------------------
export class TestAsserter extends PharoObject {
  setUp() {
  }
  tearDown() {
  }
  assert_equals_(left, right) {
    var lvalue = null;
    var rvalue = null;
    if (typeof left == 'function') {
      lvalue = left();
    } else {
      lvalue = left;
    }
    if (typeof right == 'function') {
      rvalue = right();
    } else {
      rvalue = right;
    }
    if (lvalue != rvalue)
      throw `Assertion failed: ${left} != ${right}.`;
  }

  static suiteClass() {
    return TestSuite;
  }
}

//-----------------------------------------------------------------------------
class TestResult {
  selectorName = null;
  result = null;
  message = null;

  run(proc) {
    var oldLog = console.log;
    var outer = this;
    this.message = "";
    console.log = function(message){
      outer.message = outer.message + message + "\r\n";
    };
    try {
      proc();
    } catch (e) {
      this.result = false;
      this.message = e.message;
      return;
    } finally {
      console.log = oldLog;
    }
    this.result = true;
  }
}

//-----------------------------------------------------------------------------
class TestResultSummary {
  className = null;
  eachTest = null;
  succeeded = 0;
  failed = 0;
  constructor() {
    this.eachTest = [];
  }
  push(result) {
    this.eachTest.push(result);
    if(result.result) {
      this.succeeded++;
    } else {
      this.failed++;
    }
  }
}


//-----------------------------------------------------------------------------
export class TestSuite {
  static named_(name) {
    return new TestSuite(name)
  }

  className = null;
  tests = null;

  constructor(name){
    this.tests = [];
    this.className = name;
  }

  addTest_(aTest) {
    this.tests.push(aTest);
    return this;
  }

  run() {
    var results = new TestResultSummary();
    this.tests.forEach(function(element) {
      var result = new TestResult();
      result.selectorName = element.testSelector;
      result.run(function() {
        element.setUp();
        element[element.testSelector]();
        element.tearDown();
      });
      results.push(result);
    });
    console.log(JSON.stringify(results));
  }
}


//-----------------------------------------------------------------------------
import stream from 'memory-streams';

String.cr = function() { return '\r'; };
String.lf = function() { return '\n'; };
String.crlf = function() { return '\r\n'; };
String.prototype.readStream = function() {
  return new stream.ReadableStream(this);
};
String.streamContents_ = function(aBlock) {
  var s = new stream.WritableStream();
  aBlock(s);
  return s.toString();
};
String.prototype.trimLeft_ = function(aBlock) {
  for(var i = 0; i < this.length; i++) {
    if (!aBlock(this.charAt(i))) {
      return this.substring(i);
    }
  }
  return this;
};
String.prototype.asString = function() {
  return this;
};

//-----------------------------------------------------------------------------
Error.prototype.getMessage = function() {
  return this.message;
};

Error.prototype._toString = Error.prototype.toString;
Error.prototype.toString = function() {
  return this.stack.toString();
};

//-----------------------------------------------------------------------------
import fs from 'fs';
import path from 'path';
export class FileLocator {
  constructor(path) {
    this._payload = path;
  }
  isDirectory() {
    return fs.statSync(this._payload).isDirectory();
  }
  exists() {
    return fs.existsSync(this._payload);
  }
  fullName() {
    return this._payload;
  }
  basename() {
    return path.basename(this._payload);
  }
  creationTime() {
    return fs.statSync(this._payload).ctime;
  }
  accessTime() {
    return fs.statSync(this._payload).atime;
  }
  changeTime() {
    return fs.statSync(this._payload).mtime.toLocaleString();
  }
  relativeTo_(relBase) {
    return new FileLocator(path.relative(this._payload, relBase.fullName()));
  }
  contains_(p) {
    var fullName = p;
    if (fullName.constructor.name === 'FileLocator'){
      fullName = fullName.fullName();
    }
    var abs = path.resolve(this._payload, fullName);
    return abs.startsWith(this._payload);
  }
  parent() {
    return new FileLocator(path.dirname(this._payload));
  }
  ensureCreateDirectory() {
    var dir = this.isDirectory ?
      dir = this._payload : path.dirname(this._payload);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, {recursive: true});
    }
  }
  joinPath_(p) {
    return new FileLocator(path.resolve(this._payload, p));
  }
  writeStreamDo_(aBlock) {
    var io = new SWFile(this._payload);
    try {
      aBlock(io);
    } finally {
      io.destroy();
    }
  }

  children() {
    var self = this;
    return fs.readdirSync(this._payload).flatMap(function(name) {
      return self.join(name);
    });
  }
}

FileLocator.imageDirectory = function() {
  return new FileLocator(process.cwd());
}

//-----------------------------------------------------------------------------
Date.prototype.prettyPrint = function() {
  return this.toLocaleString();
};

//-----------------------------------------------------------------------------
export class SWFile {
  constructor(path) {
    this._path = path;
  }

  truncate() {
  }

  position_(p) {
  }

  write_(contents) {
    fs.writeFileSync(this._path, contents);
  }

  destroy() {
  }
}

//-----------------------------------------------------------------------------
export class Dictionary {}

// for Dummy
export class SWGFortranCompilerAdapter {}


//-----------------------------------------------------------------------------
if (process.env.NODE_ENV == 'test') {

  var fl = FileLocator.imageDirectory();
  console.log(fl.fullName());
  console.log(fl.isDirectory());
  console.log(fl.exists());
  console.log(fl.basename());
  console.log(fl.creationTime());
  console.log(fl.accessTime());
  console.log(fl.changeTime());
  console.log(fl.relativeTo_(new FileLocator('/bin')).fullName());
  var fl2 = fl.join('hoge/fuga.txt');
  console.log(fl2.fullName());
  console.log(fl.contains_(fl2));
  fl2.ensureCreateDirectory();
  fl2.writeStreamDo_(function(io) {
    console.log(io._path);
    io.write_('Hello, World\n');
  });
  console.log(fl.children());
}
